### PR TITLE
Deprecate realm names with a leading underscore (#73366)

### PR DESCRIPTION
**Pull request by ywangd**
Deprecation warning is now issued if any realm is configured with a name
prefixed with an underscore. This applies to all realms regardless
whether they are enabled or not.

Relates: #73250

### CODE REVIEW

**Comment by tvernum** [comment]
I think we should deprecate in both `master` and `7.x`
Given the impact of forcing admins to rename realms, we may want to leave it deprecated during 8.x rather than removing support entirely.

File: docs/reference/migration/migrate_7_14.asciidoc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+[%collapsible]
+====
+*Details* +
+Elasticsearch creates "synthetic" realm names on the fly for services like API keys.
+These synthetic realm names are prefixed with an underscore.
+Currently, user configured realms can also be given a name with a leading underscore.
+This creates confusion since realm names are meant to be unique for a node.
+
+*Impact* +
+Configuring a realm name with a leading underscore is deprecated. In 8.0.0,
```
Comment (tvernum) [inline docs/reference/migration/migrate_7_14.asciidoc:66]: I don't think we should document that things _will_ happen in 8.0 if we haven't actually planned to do that work.
Perhaps,
```suggestion
Configuring a realm name with a leading underscore is deprecated. In a future releases of {es}
```

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java [hunk 2]
```diff
@@ -167,6 +170,36 @@ static DeprecationIssue checkImplicitlyDisabledBasicRealms(final Settings settin
...
+            if (realmIdentifier.getName().startsWith(RESERVED_REALM_NAME_PREFIX)) {
+                reservedPrefixedRealmIdentifiers.add(realmIdentifier);
+            }
+        }
+        if (reservedPrefixedRealmIdentifiers.isEmpty()) {
+            return null;
+        } else {
+            return new DeprecationIssue(
+                DeprecationIssue.Level.CRITICAL,
+                "Realm names cannot start with [" + RESERVED_REALM_NAME_PREFIX + "] in next major release.",
```
Comment (tvernum) [inline x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java:189]: Per above comments, we need to decide on our plans for 8.0 in order to work out what message to include here.

Comment (ywangd) [reply x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java]: I changed all places to say "a future major release".

**Comment by ywangd** [comment]
> I think we should deprecate in both `master` and `7.x`
> Given the impact of forcing admins to rename realms, we may want to leave it deprecated during 8.x rather than removing support entirely.

Makse sense. I raised #73366 for deprecation in 8.0

**Pull request merged**